File: year2023/day02/typescript/solution.ts
import { readFileSync } from 'fs';

const file: any = readFileSync('../' + (process.argv[2] === 'puzzle' ? 'puzzle' : 'example') + '.txt', 'utf-8');
const input: string[] = file.toString().trim().split('\n');
const max: Record<string, number> = { red: 12, green: 13, blue: 14 };
if(process.argv[2] === 'puzzle') {
  console.log(part_1());
  console.log(part_2());
}

export function part_1(): number {
  return input.reduce((acc, row) => {
    return row.split(": ")[1].split(/; /g).some((subs) => subs.split(/, /g).some((tuple) => {
      return Number(tuple.trim().split(" ")[0]) > max[tuple.trim().split(" ")[1]];
    })) ? acc : acc + Number(row.split(": ")[0].split(" ")[1]); 
  }, 0);
}

export function part_2(): number {
  return input.reduce((acc, row) => {
    const map: Map<string, number> = new Map();
    row.split(": ")[1].split(/; /g).forEach(subs => subs.split(/, /g).forEach(tuple => map.set(tuple.trim().split(" ")[1], Math.max(map.get(tuple.trim().split(" ")[1]) ?? 0, Number(tuple.trim().split(" ")[0])))));
    return acc + [...map.values()].reduce((acc, value) => acc * value, 1);
  }, 0);
}

File: year2023/day02/python/solution.py
def main():
  print(part_1())
  print(part_2())


def part_1():
  color_dict = { 'red': 12, 'green': 13, 'blue': 14 }
  return sum(int(row[1]) for row in input if all(int(row[i]) <= color_dict[row[i + 1]] for i in range(2, len(row), 2)))


def part_2():
  sum = 0
  for row in input:
    power_dict = {'red': 0,'green': 0,'blue': 0}

    for i in range(2, len(row), 2):
      if int(row[i]) > power_dict[row[i + 1]]:
        power_dict[row[i + 1]] = int(row[i])

    sum += (power_dict['red'] * power_dict['green'] * power_dict['blue'])

  return sum


def setup(path):
  global input

  with open(path, 'r') as file:
    input = [row.replace(':', '').replace(',', '').replace(';', '').split() for row in file.read().split('\n')]


if __name__ == '__main__':
  setup('../puzzle.txt')
  main()